backend/src/postsDataSource.ts

```diff
@@ -15,36 +18,63 @@ export class PostsDataSource extends DataSource {
     return this.users
   }
 
+  existsUser (id) {
+    if (this.users.find(user => user.id === id)) {
+      return true
+    } else {
+      return false
+    }
+  }
+
   getPosts () {
     return this.posts
   }
 
-  upvote (postId, voter) {
+  upvote (postId, userId) {
     const foundItem = this.getPosts().find(item => item.id === postId)
-    const foundUser = this.getUsers().find(user => user.name === voter.name)
+    const foundUser = this.getUsers().find(user => user.id === userId)
     if (foundItem && foundUser) {
-      if (!foundItem.voters.includes(foundUser.name)) {
+      if (!foundItem.voters.includes(foundUser.id)) {
         foundItem.votes += 1
-        foundItem.voters.push(foundUser.name)
+        foundItem.voters.push(foundUser.id)
       }
     }
     return foundItem
   }
 
-  addPost (newPost) {
-    const foundUser = this.getUsers().find(user => user.name === newPost.author.name)
+  addPost (newPost, userId) {
+    const foundUser = this.getUsers().find(user => user.id === userId)
     if (foundUser) {
       const item = { id: uuidv4(), title: newPost.title, votes: 0, voters: [], author: foundUser }
       foundUser.posts.push(item)
       this.posts.push(item)
       return item
+    }
+    return null
+  }
+
+  login (email, password) {
+    const foundUser = this.getUsers().find(user => user.email === email)
+
+    if (foundUser) {
+      if (bcrypt.compareSync(password, foundUser.password)) {
+        return jwt.sign({ id: foundUser.id }, process.env.JWTSECRET, { algorithm: 'HS256' })
+      }
+    }
+  }
+
+  addUser (name, email, password) {
+    if (password.length >= 8) {
```

Comment:
Prefer guard-clauses over nested if-clauses, for readability:

```suggestion
    if (password.length < 8) throw new UserInputError('Password must have at least 8 characters')
```